### ISSUE TITLE
Instructions on how to cleanly handle streams and connections

### ISSUE THREAD

**Issue by MatthewLymer** [Matthias247/http2dotnet#3]
I am using the http2dotnet library to implement a mitm proxy for h2 traffic.  I am using chrome to visit https://www.httpvshttps.com/ to do simple load testing.  (I'm using dotnet core 2.1-preview for ALPN support)

So far, I am able to view that page a few times, the first couple times the performance is acceptable (roughly 1/10th the speed of a direct connection), and then the performance plummets to about 1/100th, and then finally it just freezes my console app entirely.

This leads to me to believe that I am not disposing of something properly, but from the examples I cannot determine what I am missing, aside from disposing of the `IStream` implementations.

My code looks as follows:

```csharp
private async Task HandleHttp2TrafficAsync(
  Stream clientStream, // proxy client
  Stream serverStream, // remote server
  CancellationToken cancellationToken)
{
  var clientConfiguration = new ConnectionConfigurationBuilder(false)
    .UseHuffmanStrategy(HuffmanStrategy.IfSmaller)
    .Build();

  var serverStreams = serverStream.CreateStreams();

  var clientConnection = new Connection(
    clientConfiguration,
    serverStreams.ReadableStream,
    serverStreams.WriteableStream);

  var serverConfiguration = new ConnectionConfigurationBuilder(true)
    .UseStreamListener(stream =>
    {
      OnStreamAsync(
	    stream, 
		clientConnection.CreateStreamAsync, 
		cancellationToken)
      .ContinueWith(tt => { }, cancellationToken);

      return true;
    })
    .Build();

  var clientStreams = clientStream.CreateStreams();

  var serverConnection = new Connection(
    serverConfiguration,
    clientStreams.ReadableStream,
    clientStreams.WriteableStream);

  await Task.WhenAll(serverConnection.Done, clientConnection.Done);
}

private static async Task OnStreamAsync(
  IStream serverStream, 
  Func<IEnumerable<HeaderField>, bool, Task<IStream>> createClientStreamAsync, 
  CancellationToken cancellationToken)
{
  var buffer = ArrayPool<byte>.Shared.Rent(8 * 1024);

  try
  {
    // start reading the request
    var requestHeaders = await serverStream.ReadHeadersAsync();

    using (var requestBodyStream = new MemoryStream())
    {
      while (true)
      {
        var result = await serverStream.ReadAsync(buffer);

        requestBodyStream.Write(buffer, 0, result.BytesRead);

        if (result.EndOfStream)
        {
          break;
        }
      }
      
      // start sending the request
      using (var clientStream = await createClientStreamAsync(
	    requestHeaders, 
		requestBodyStream.Length == 0))
      {
        if (requestBodyStream.Length > 0)
        {
          await clientStream.WriteAsync(requestBodyStream.ToArray(), true);
        }

        // start reading the response
        var responseHeaders = await clientStream.ReadHeadersAsync();

        using (var responseBodyStream = new MemoryStream())
        {
          while (true)
          {
            var result = await clientStream.ReadAsync(buffer);

            responseBodyStream.Write(buffer, 0, result.BytesRead);

            if (result.EndOfStream)
            {
              break;
            }

            // start sending the response
            await serverStream.WriteHeadersAsync(
			  responseHeaders, 
			  responseBodyStream.Length == 0); // line 351

            if (responseBodyStream.Length > 0)
            {
              await serverStream.WriteAsync(
			    responseBodyStream.ToArray(), 
				true);
            }
          }
        }
      }
    }
  }
  finally
  {
    serverStream.Dispose(); // should I dispose this?
    ArrayPool<byte>.Shared.Return(buffer);
  }
}
```

And my general run through I receive the following exceptions after setting up a global exception handler

```
      An unhandled exception occured
System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Attempted to write headers after data) ---> System.Exception: Attempted to write headers after data
   at Http2.StreamImpl.<WriteValidatedHeadersAsync>d__31.MoveNext()
   at MyApp.Proxy.Core.Http.HttpSocketProcessor.<OnStreamAsync>d__12.MoveNext() in C:\Projects\myapp-proxy\src\MyApp.Proxy.Core\Http\HttpSocketProcessor.cs:line 351
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.Exception: Attempted to write headers after data
   at Http2.StreamImpl.<WriteValidatedHeadersAsync>d__31.MoveNext()
   at MyApp.Proxy.Core.Http.HttpSocketProcessor.<OnStreamAsync>d__12.MoveNext() in C:\Projects\myapp-proxy\src\MyApp.Proxy.Core\Http\HttpSocketProcessor.cs:line 351<---
```

```
      An unhandled exception occured
System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Attempted to read past the end of the stream.) ---> System.IO.EndOfStreamException: Attempted to read past the end of the stream.
   at Http2.ByteStreamExtensions.<ReadAll>d__0.MoveNext()
   at Http2.ClientPreface.<ReadAsync>d__5.MoveNext()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.IO.EndOfStreamException: Attempted to read past the end of the stream.
   at Http2.ByteStreamExtensions.<ReadAll>d__0.MoveNext()
   at Http2.ClientPreface.<ReadAsync>d__5.MoveNext()<---
```

```
      An unhandled exception occured
System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (One or more errors occurred. (Unable to transfer data on the transport connection: The I/O operation has been aborted because of either a thread exit or an application request.)) ---> System.AggregateException: One or more errors occurred. (Unable to transfer data on the transport connection: The I/O operation has been aborted because of either a thread exit or an application request.) ---> System.IO.IOException: Unable to transfer data on the transport connection: The I/O operation has been aborted because of either a thread exit or an application request. ---> System.Net.Sockets.SocketException: The I/O operation has been aborted because of either a thread exit or an application request
   --- End of inner exception stack trace ---
   at System.Net.Security.SslStreamInternal.<<FillBufferAsync>g__InternalFillBufferAsync|36_0>d`1.MoveNext()
   at System.Net.Security.SslStreamInternal.<ReadAsyncInternal>d__32`1.MoveNext()
   --- End of inner exception stack trace ---
   at Http2.IoStreamExtensions.IoStreamWrapper.<>c.<ReadAsync>b__2_0(Task`1 tt)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at Http2.ByteStreamExtensions.<ReadAll>d__0.MoveNext()
   at Http2.ClientPreface.<ReadAsync>d__5.MoveNext()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.AggregateException: One or more errors occurred. (Unable to transfer data on the transport connection: The I/O operation has been aborted because of either a thread exit or an application request.) ---> System.IO.IOException: Unable to transfer data on the transport connection: The I/O operation has been aborted because of either a thread exit or an application request. ---> System.Net.Sockets.SocketException: The I/O operation has been aborted because of either a thread exit or an application request
   --- End of inner exception stack trace ---
   at System.Net.Security.SslStreamInternal.<<FillBufferAsync>g__InternalFillBufferAsync|36_0>d`1.MoveNext()
   at System.Net.Security.SslStreamInternal.<ReadAsyncInternal>d__32`1.MoveNext()
   --- End of inner exception stack trace ---
   at Http2.IoStreamExtensions.IoStreamWrapper.<>c.<ReadAsync>b__2_0(Task`1 tt)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at Http2.ByteStreamExtensions.<ReadAll>d__0.MoveNext()
   at Http2.ClientPreface.<ReadAsync>d__5.MoveNext()
---> (Inner Exception #0) System.IO.IOException: Unable to transfer data on the transport connection: The I/O operation has been aborted because of either a thread exit or an application request. ---> System.Net.Sockets.SocketException: The I/O operation has been aborted because of either a thread exit or an application request
   --- End of inner exception stack trace ---
   at System.Net.Security.SslStreamInternal.<<FillBufferAsync>g__InternalFillBufferAsync|36_0>d`1.MoveNext()
   at System.Net.Security.SslStreamInternal.<ReadAsyncInternal>d__32`1.MoveNext()<---
```

Sorry for the whole rig-ma-role of an 'issue', I just need a pointer on how I'm supposed to manage the connections and streams.

Also, I noticed that the API does not accept `CancellationToken`s, is there a pattern that is expected for stopping connections and streams (though I do notice that `IStream` has a `Cancel` method).

Thank you for your help and guidance,
Matthew

**Comment by Matthias247** [comment]
There are two things in HTTP/2: Streams and Connections. Between two peers there is always a connection, which runs on top of TCP/TLS connection, and where HTTP/2 specific frames are exchanged. On top of that clients can create multiple streams, where each stream represents a single HTTP request. These concepts are represented here in the library.

So there are also two things which can be closed, the stream and the connection.
Streams are automatically closed when the HTTP request had been fully finished. If that's not the case, and e.g. one part of the connection broke, the `.Cancel()` method can be used to close/cancel the stream. This is equivalent to the `.Dispose()` method. It is safe to Cancel finished requests, to it can be called in a finally block.

Then there is the connection. This one should get automatically closed if the remote side closes the underlying TCP connection. In case you want to do it or the other side doesn't do it, there are multiple  multiple APIs on the `Connection` object:
- CloseNow()
- GoAwayAsync()

The doc comments should carry the relevant details for those.

Hope that helps

**Comment by Matthias247** [comment]
Edit: The `Dispose` method on the stream forwards to `Cancel()`. So it would be the right thing to add this line. It will mainly have an effect when one of the sides breaks down without completing the request. But since this will be common on a server it's good.

Apart from that the code looks pretty reasonable.

You can most likely improve performance by not reading the complete request into a MemoryStream, but directly forwarding the chunks. This would improve latency as well as decrease memory consumption. It is likely that the allocations of those MemoryStreams and the garbage collector hurt the performance.

And btw: Full duplex forwarding should also be possible. It's not necessary to wait until the the request body was sent before the response can be read and forwarded. Would be necessary to proxy things like grpc bidirectional streaming.

**Comment by MatthewLymer** [comment]
Thank you for the great info, I appreciate the time you spent making the library and replying to my issue.

The `MemoryStream` implementation is just to get this thing working, I'll have more of a streaming implementation when I better understand the flow, the GC is going wild with my implementation, I'm just in the learning phase right now :).

I'll post a more complete example once I have implemented your suggestions, in case anyone else is trying what I'm doing.

**Comment by Matthias247** [comment]
I'm closing this since there was no more activity. If there is any issue on the library, feel free to reopen or create a new issue.
